cactus.rb

```diff
@@ -143,6 +85,17 @@ def install
 
   end
 
+  private def extract_python_dependencies(libexec)
+    wrapper = 'from distutils.core import run_setup;
+    result = run_setup("setup.py", stop_after="init");
+    f = open("toil-requirement.txt","a");
+    f.write("\n".join(result.install_requires));
+    f.close();'
+    File.write('toil-requirement.py', wrapper.split(/\s*;\s*/).join("\n"), mode: "a")
+    system "#{libexec}/bin/python3",
+    	"toil-requirement.py"
+  end
+
```

Comment:
To simplify the creation of a formula, I end up using this function to extract Cactus dependencies from the `setup.py` file (`install_requires` list) to amend them in the `toil-requirement.py` file. By doing so, python dependencies are installed using pip from virtualenv. However, the instalment of python dependencies using this approach instead of Formula resources tag is not a recommended approach by Formula Cookbook.

Reply:
@muffato Let me know if you agree with this approach please =)

Reply:
What's the benefit of doing that ? To simplify the formula ?

Reply:
Actually two benefits:
1 - Simplification to create a formula; no need to specify each dependency as a resource (with link and SHA1)
2 - The dependencies of a dependency is not installed by the Formula using the resource tag, which is a nightmare to figure out all the dependencies of a dependency.

Reply:
I see, it's a compromise :) It's not the advised best practice, but it makes the formula much simpler.
Would it have been possible to use this method to get the complete (recursive) list of dependencies on your laptop, and then use brew's tool to import them as a complete list of `resource` in the formula ? That's >100 dependencies according to my count, and I think it would horribly clutter the formula, but just checking for my understanding.

I'm happy either way, what you have now is simpler and more maintainable :+1: but I would suggest to still mention your previous implementation (`resource`) in Confluence

Reply:
If I understand correctly, they use the existing functionality to create a virtualenv, and then run their own pip & others commands in that virtualenv ? Here, I guess you could directly run `setup.py` with the virtualenv's python and you wouldn't have to collect its requirements into a file ?

Reply:
Exactly. But the run of `setup.py` directly does not give the link step for "free".  In the end,  we have two options:

1.)  It is the way as the formula is implemented. The `venv.pip_install_and_link buildpath` install Cactus as a python dependency and link the executables from `#{prefix}/libexec/bin` to `#{prefix}/bin` for free. However, this option uses `pip` call with `--no-deps` flag, and thus the dependencies should be collected for instalment in another `pip` call.

2.) We directly run `setup.py` within the virtualenv's python and everything is installed in one `pip` call. However, symbolic links must be created manually. 

We just need to pick an approach one of these two approaches.

Reply:
Yep, I agree. Option 2 looks better. So you would sort of embed the content of `pip_install_and_link`/`pip_install` into the cactus formula, but tailored to just call `setup.py` with the virtualenv's `pip` instead of calling the default `do_install` ?